semaphore.hpp

```diff
@@ -0,0 +1,429 @@
+                                                                        oldCount + 1,
+                                                                        std::memory_order_relaxed,
+                                                                        std::memory_order_relaxed))
+                        return 0;
+                }
+            }
+        }
+        if (max > 1)
+            return 1 + tryWaitMany(max - 1);
+        return 1;
+    }
+
+public:
+    LightweightSemaphore(ssize_t initialCount = 0) : m_count(initialCount)
+    {
+        assert(initialCount >= 0);
+    }
+
+    bool tryWait()
+    {
+        ssize_t oldCount = m_count.load(std::memory_order_relaxed);
+        while (oldCount > 0)
+        {
+            if (m_count.compare_exchange_weak(oldCount,
+                                              oldCount - 1,
+                                              std::memory_order_acquire,
+                                              std::memory_order_relaxed))
+                return true;
+        }
+        return false;
+    }
+
+    void wait()
+    {
+        if (!tryWait())
+            waitWithPartialSpinning();
+    }
+
+    bool wait(std::int64_t timeout_usecs)
+    {
+        return tryWait() || waitWithPartialSpinning(timeout_usecs);
+    }
+
+    // Acquires between 0 and (greedily) max, inclusive
+    ssize_t tryWaitMany(ssize_t max)
+    {
+        assert(max >= 0);
+        ssize_t oldCount = m_count.load(std::memory_order_relaxed);
+        while (oldCount > 0)
+        {
+            ssize_t newCount = oldCount > max ? oldCount - max : 0;
+            if (m_count.compare_exchange_weak(oldCount, newCount, std::memory_order_acquire, std::memory_order_relaxed))
+                return oldCount - newCount;
+        }
+        return 0;
+    }
+
+    // Acquires at least one, and (greedily) at most max
+    ssize_t waitMany(ssize_t max, std::int64_t timeout_usecs)
+    {
+        assert(max >= 0);
+        ssize_t result = tryWaitMany(max);
+        if (result == 0 && max > 0)
+            result = waitManyWithPartialSpinning(max, timeout_usecs);
+        return result;
+    }
+
+    ssize_t waitMany(ssize_t max)
+    {
+        ssize_t result = waitMany(max, -1);
+        assert(result > 0);
+        return result;
+    }
+
+    void signal(ssize_t count = 1)
+    {
+        assert(count >= 0);
+        ssize_t oldCount = m_count.fetch_add(count, std::memory_order_release);
+        ssize_t toRelease = -oldCount < count ? -oldCount : count;
+        if (toRelease > 0)
+        {
+            m_sema.signal((int) toRelease);
+        }
+    }
+
+    ssize_t availableApprox() const
+    {
+        ssize_t count = m_count.load(std::memory_order_relaxed);
+        return count > 0 ? count : 0;
+    }
+};
+}   // end namespace mpmc_sema
+}   // end namespace details
+
+typedef details::mpmc_sema::LightweightSemaphore LightweightSemaphore;
```

Comment:
This will cause problems if the semaphore included with my `ReaderWriterQueue` takes the same approach, since they're both named `LightweightSemaphore`. Perhaps that one should be renamed or moved to another namespace, since it's less general.

I also don't like putting it in a namespace and then having a typedef -- I'd prefer just putting it in the right namespace to start with. Backwards compatibility is a non-issue here since the code was previously in the `details` internal namespace.